services/harmony/app/util/cache/memory-cache.ts

```diff
@@ -4,30 +4,65 @@ import env from '../env';
 
 // Simple implementation of a string cache backed by an in-memory
 // least-recently-used (LRU) cache
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type FetchMethod = (key: string, context?: any) => Promise<string>;
+
+interface MemoryCacheOptions {
+  ttl?: number; // Optional TTL in milliseconds
+  maxSize?: number; // Optional max size override
+}
+
 export class MemoryCache extends Cache {
-  data: LRUCache<string, string>;
+  private data: LRUCache<string, string>;
+
+  private fetchMethod: FetchMethod;
+
+  private pending: Map<string, Promise<string>> = new Map();
 
-  constructor(fetchMethod: (k: string) => Promise<string>) {
+  constructor(fetchMethod: FetchMethod, options?: MemoryCacheOptions) {
     super();
-    const options = {
-      maxSize: env.maxDataOperationCacheSize,
-      sizeCalculation: (v: string): number => {
-        return v.length;
-      },
-      fetchMethod,
-    };
-    this.data = new LRUCache(options);
+    this.fetchMethod = fetchMethod;
+
+    this.data = new LRUCache<string, string>({
+      maxSize: options?.maxSize ?? env.maxDataOperationCacheSize,
+      ttl: options?.ttl, // No expiration if undefined
+      sizeCalculation: (value: string): number => value.length,
+    });
   }
 
-  async get(key: string): Promise<string> {
+  async get(key: string): Promise<string | undefined> {
     return this.data.get(key);
   }
 
-  async fetch(key: string): Promise<string> {
-    return this.data.fetch(key);
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async fetch(key: string, context?: any): Promise<string> {
+    const cached = this.data.get(key);
+    if (cached !== undefined) return cached;
+
+    // Check for in-progress fetch
+    const existingPromise = this.pending.get(key);
```

Comment:
We should add a unit test to verify that if the key is in the pending list the existing promise is returned and a new one is not created.

Reply:
I added a unit test.